### PR TITLE
Make reshape and view on Memory produce Arrays and delete wrap

### LINKED ISSUE #53552: `wrap` is a very generic name to export
`wrap` was introduced as the safe alternative to `unsafe_wrap` in https://github.com/JuliaLang/julia/pull/52049 in December by @MasonProtter. 

Maybe this is the right call but I want to at least flag the issue. `wrap` a very a generic name and there are many packages using it for all kinds of purposes, and many others that will likely want to choose that name in the future. [288 files](https://github.com/search?q=%22function+wrap%28%22+language%3Ajulia&type=code) on GitHub define `wrap`, with some duplicates.

Options:
- No change. Users will qualify `Package.wrap` instead of `using` packages. Many people who would write a function `wrap` will probably decide to choose a different name for their function. Heavy users of `Base.wrap` will be pleased.
- Don't export `wrap`. Users will write `Base.wrap` instead of `wrap`, which could be annoying if it's in common usage, but will be more pleasant for authors and users of other functions named `wrap`.
- Rename it. `wrap_memory`? `Array`? `array`? `memoryarray`? something else?

### PR DESCRIPTION
**Pull request by LilithHafner**
Make reshape and (one-based) view on Memory produce Arrays and delete wrap

Implements https://github.com/JuliaLang/julia/issues/53552#issuecomment-2024288283

Fixes #53552

### CODE REVIEW

File: base/genericmemory.jl
Comment (LilithHafner) [inline base/genericmemory.jl:302]: @oscardssmith, I think this `@inbounds` is not preventing bounds checking which causes the build to fail (even though evaling these into Base in a pre-built Julia does work) . Is there a more robust way to disable bounds checking or should we avoid constructing this out of bounds `MemoryRef` entirely and branch on `isempty(inds)`?

Comment (vtjnash) [reply base/genericmemory.jl:302]: Since first!=1 in general, you need the branch on isempty anyways to handle that edge case

Comment (vtjnash) [reply base/genericmemory.jl:302]: `@inbounds` is undefined behavior if it would have thrown an error when disabled. That can lead to arbitrary memory corruption and program corruption.

Comment (LilithHafner) [reply base/genericmemory.jl:302]: Is it possible to create an out of bounds MemoryRef at an offset, or is that always UB?

Comment (vtjnash) [reply base/genericmemory.jl:302]: It is considered UB to make a MemoryRef that does not reference valid memory

**Comment by N5N3** [comment]
The `reshape` change seems ok as `reshape` of `Array` returns `Array`.
But I'm afraid `view` is another story, as it seems incompatible with offset case.

File: base/genericmemory.jl
Comment (LilithHafner) [reply base/genericmemory.jl:302]: This makes sense. And what makes `MemoryRef(Memory{Int64}(undef, 0))` is okay is

https://github.com/JuliaLang/julia/blob/313f933370aff4edf1eb8f972964b2e459451066/src/builtins.c#L1678-L1679

Comment (vtjnash) [reply base/genericmemory.jl:302]: Yes, it both does (technically) point to valid memory (size 0 memory allocate extra space for 1 element) and is special cases to be allowed for the zero-sized case to be constructed, but not computed with it

Comment (oscardssmith) [reply base/genericmemory.jl]: possibly too late, but would the design have been simpler if we just disallowed 0 length memory? since it's not resizable, there isn't much need to be able to make one.

**Comment by LilithHafner** [comment]
I dropped the `@inbounds` and restricted `view` to `Base.OneTo` and `UnitRange` inds. If there were an AbstractOneBasedUintRange type or an isonebased trait on abstract array types I would have used that instead.

Should we support slices?
```julia
view(m::GenericMemory{M, T}, inds::Colon) where {M, T} = view(m, eachindex(m))
```

File: base/genericmemory.jl
Comment (LilithHafner) [reply base/genericmemory.jl]: The fact that empty memories are actually stored as length 1 memories supports that. Empty arrays can still be backed by the "one true size-one memory" of that type to avoid allocations.

Comment (LilithHafner) [reply base/genericmemory.jl]: Moving this discussion to https://github.com/JuliaLang/julia/issues/53903#issue-2216508204

File: base/iobuffer.jl
Comment (cafaxo) [inline base/iobuffer.jl:459]: This seems to produce a `view` of length `nbytes+1`. Is that intended?

File: base/iobuffer.jl
Comment (cafaxo) [inline base/iobuffer.jl:468]: Same issue here.

File: base/iobuffer.jl
Comment (LilithHafner) [reply base/iobuffer.jl:459]:
```suggestion
            data = view(io.data, io.offset+1:nbytes+io.offset)
```

File: base/iobuffer.jl
Comment (LilithHafner) [reply base/iobuffer.jl:468]:
```suggestion
            data = view(io.data, io.ptr:io.ptr+nbytes-1)
```

File: base/iobuffer.jl
Comment (LilithHafner) [reply base/iobuffer.jl]: Thank you. That was not intended.

File: base/genericmemory.jl
Comment (LilithHafner) [reply base/genericmemory.jl]: There is no long `@inbounds` in the current implementation.

File: base/genericmemory.jl
Comment (vtjnash) [inline base/genericmemory.jl:301]: Why doesn't it matter that OneTo can be empty?
```suggestion
    isempty(inds) && return T[] # needed to allow view(Memory{T}(undef, 0), 2:1)
```

You can always construct `MemoryRef(m)`, but can't always then offset it by `MemoryRef(ref, first(inds))`

**Comment by vtjnash** [comment]
> But I'm afraid view is another story, as it seems incompatible with offset inds.

I think `reshape(view(a, n:m), (x, y))` should be possible for the compiler to be exactly as optimized?

**Review by vtjnash — APPROVED** [review]

File: base/genericmemory.jl
Comment (LilithHafner) [reply base/genericmemory.jl:301]: I was thinking of the following implementation which I believe is safe and should be faster:

```julia
@eval function view(m::GenericMemory{M, T}, inds::UnitRange) where {M, T}
    isempty(inds) && return T[] # needed to allow view(Memory{T}(undef, 0), 2:1)
    @boundscheck checkbounds(m, inds)
    ref = @inbounds MemoryRef(m, first(inds)) # @inbounds here is safe because of the checkbounds and isempty checks above
    dims = (length(inds),)
    $(Expr(:new, :(Array{T, 1}), :ref, :dims))
end
@eval function view(m::GenericMemory{M, T}, inds::OneTo) where {M, T}
    @boundscheck checkbounds(m, inds)
    ref = MemoryRef(m) # Does not perform bounds checking, but empty memory gets special handling to make this okay
    dims = (length(inds),)
    $(Expr(:new, :(Array{T, 1}), :ref, :dims))
end
```

But that implementation doesn't benchmark as faster for me so I'll just go with the simpler one in this PR with more checks.

**Comment by LilithHafner** [comment]
I think @N5N3 was referring to offset indices like `OffsetArrays.IdOffsetRange(values=3:6, indices=53:56)`, and they were correct that the proposed implementation was buggy for those offset inds. Fixed in 91542bc7a12e8f28f665aa16cad829818971df36.

**Comment by LilithHafner** [comment]
IIUC it is possible for resizing a vector produced here to corrupt the underlying memory. If I'm correct, that should be documented.

**Comment by LilithHafner** [comment]
Seemingly unrelated CI failure, but master is green. Updating branch.

**Comment by LilithHafner** [comment]
The failure, `all(T -> T <: AbstractString, Base.return_types(string))`, was real. IDK if that's a valid or useful thing to test, but I fixed it by telling the compiler that `StringVector` will indeed return a `Vector{UInt8}`.

**Comment by oscardssmith** [comment]
I think it might be an issue if the compiler isn't able to infer the type of taking a view of a Memory.

**Comment by LilithHafner** [comment]
I'm ready to merge if @vtjnash and @oscardssmith think it's ready.

**Review by oscardssmith — APPROVED** [review]

**Comment by LilithHafner** [comment]
Nothing complex or controversial has changed since @vtjnash's approval. I'm going to go ahead and merge.

File: base/iobuffer.jl
Comment (odow) [inline base/iobuffer.jl:45]: I think this change broke JSON3:
```Julia
(base) oscar@Oscars-MBP JSONSchema % julia +nightly
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.12.0-DEV.313 (2024-04-07)
 _/ |\__'_|_|_|\__'_|  |  Commit c5a3b653353 (0 days old master)
|__/                   |

julia> Base.StringVector(UInt64(2))
ERROR: TypeError: in new, expected Tuple{Int64}, got a value of type Tuple{UInt64}
Stacktrace:
 [1] view
   @ ./genericmemory.jl:312 [inlined]
 [2] StringVector(n::UInt64)
   @ Base ./iobuffer.jl:45
 [3] top-level scope
   @ REPL[1]:1

julia> exit()
(base) oscar@Oscars-MBP JSONSchema % julia         
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.10.2 (2024-03-01)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> Base.StringVector(UInt64(2))
2-element Vector{UInt8}:
 0x50
 0x47
```

Example: https://github.com/quinnj/JSON3.jl/blob/6429adafd02a7ad07fa80e9a0ad21b098635244c/src/write.jl#L32

MathOptInterface failing on nightly: https://github.com/jump-dev/MathOptInterface.jl/actions/runs/8591539106/job/23540421403?pr=2464

image

Comment (LilithHafner) [reply base/iobuffer.jl:45]: Thank you for reporting. I can confirm that this is a bug and this PR introduced it.

Comment (LilithHafner) [reply base/iobuffer.jl:45]: PR fixing this: #53991

**Comment by longemen3000** [comment]
there is also a bug on CSV.jl that seems to be activated with this PR

**Comment by odow** [comment]
> there is also a bug on CSV.jl that seems to be activated with this PR

Can you elaborate? Reproducible example and stack trace?

**Comment by longemen3000** [comment]
> > there is also a bug on CSV.jl that seems to be activated with this PR
> 
> Can you elaborate? Reproducible example and stack trace?
the main problem is that `CSV.jl` depended on `Base.wrap`, that was removed in this PR. but the dependency was just to fix another bug, so i already made a PR fixing the breakage (https://github.com/JuliaData/CSV.jl/pull/1133)

**Pull request merged**
